### PR TITLE
[Core] Optimize top-k + top-p sampling by avoiding full vocabulary sort

### PR DESCRIPTION
**Pull request by nadavrot**
Summary:
This diff accelerates sampling on targets that have a slow sort implementations. When both top-k and top-p filtering are applied, the current implementation sorts the entire vocabulary (204K tokens for LLama4). This diff adds a new optimized fast-path that uses topk() to extract only the top-k elements, then sorts only those, which is faster.

  | Batch |   Vocab |    k | Baseline (ms) | Optimized (ms) | Speedup |
  |-------|---------|------|---------------|----------------|---------|
  |  1024 |  32,768 |  100 |          7.53 |           1.33 |   5.67x |
  |  1024 |  32,768 | 1000 |          7.53 |           1.55 |   4.87x |
  |  2048 | 131,072 |  100 |         57.59 |           7.70 |   7.48x |
  |  2048 | 131,072 | 1000 |         57.61 |           8.22 |   7.00x |
  |  2048 | 262,144 |  100 |        115.14 |          14.97 |   7.69x |
  |  2048 | 262,144 | 1000 |        115.09 |          15.55 |   7.40x |

 Signed-off-by: Nadav Rotem <nrotem@meta.com>

Test Plan:
Added two tests and ran the sample tests:
- test_v1_sampler
- test_v1_topk_topp_sampler
- test_kernels_top_k_per_row

Reviewed By: diviramon, wushidonguc

Differential Revision: D90454233

---

> [!NOTE]
> **Performance:** Optimizes `apply_top_k_top_p` when both `k` and `p` are set by introducing `apply_top_k_and_top_p`, which uses `topk` to extract only the top‑k logits and sorts that subset. Falls back to the general full‑sort path when any row has `k == vocab_size`.
> 
> - Adds `apply_top_k_and_top_p` to mask via partial sort and scatter back in place, preserving behavior
> - Wires optimized path behind a `max_k < vocab_size` check in `apply_top_k_top_p`
> - Tests: new correctness and equivalence tests for the optimized path; extends existing sampler tests and device fixture in `tests/v1/sample/test_topk_topp_sampler.py`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 7c66555248199fe60354f0cd839b19afcf41dc55. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Improves combined top‑k + top‑p masking performance and adds tests to validate correctness and parity with the existing path.
> 
> - Adds `apply_top_k_and_top_p` to mask by extracting `topk` logits, sorting only those, and scattering back in place; assumes `max(k) < vocab_size`
> - Wires optimized branch in `apply_top_k_top_p` guarded by `max_k < vocab_size`, otherwise falls back to full‑sort path
> - Tests: new correctness and equivalence tests for the optimized path; extends sampler test fixture and imports in `tests/v1/sample/test_topk_topp_sampler.py`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit bf7fa630d90ab4b5670c450409f9634e726c1d89. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 1]
```diff
@@ -279,6 +287,58 @@ def apply_top_k_top_p(
...
+    sorted_values, sort_indices = topk_values.sort(dim=-1, descending=False)
+
+    # Apply per-row top-k mask.
+    num_to_mask = max_k - k.to(torch.long)  # [B]
+    positions = torch.arange(max_k, device=logits.device)  # [max_k]
+    top_k_mask = positions.unsqueeze(0) < num_to_mask.unsqueeze(1)  # [B, max_k]
+    sorted_values.masked_fill_(top_k_mask, -float("inf"))
+
+    # Apply top-p on the sorted values.
+    probs_sort = sorted_values.softmax(dim=-1)
```
Comment (njhill) [inline vllm/v1/sample/ops/topk_topp_sampler.py:324]: You need to do softmax on the full logits and then gather based on `sort_indices`, or the probabilities will be incorrect.

Comment (njhill) [reply vllm/v1/sample/ops/topk_topp_sampler.py:324]: Oh sorry I guess maybe I'm wrong since in the existing impl it's applied to the already top-k masked logits. I wonder if that's actually correct/expected behaviour.

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 2]
```diff
@@ -255,6 +255,14 @@ def apply_top_k_top_p(
         # Avoid sorting vocab for top-k only case.
         return apply_top_k_only(logits, k)
 
+    if p is not None and k is not None:
+        # Use optimized path only when max_k is small enough to benefit.
+        # When k = vocab_size for any row, fall back to the general path.
+        vocab_size = logits.shape[1]
+        max_k = int(k.max().item())
```
Comment (yewentao256) [inline vllm/v1/sample/ops/topk_topp_sampler.py:262]: `item()` might cause a sync that affects performance

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 3]
```diff
@@ -279,6 +287,58 @@ def apply_top_k_top_p(
...
+    then sorts only those elements. This is much faster than sorting the
+    entire vocabulary for large vocab sizes.
+
+    The logits tensor is updated in-place.
+
+    Note: This function assumes max(k) < vocab_size. The caller should
+    fall back to the general path when any row has k = vocab_size.
+    """
+    # Get the maximum k value to determine how many elements to extract.
+    max_k = int(k.max().item())
```
Comment (yewentao256) [inline vllm/v1/sample/ops/topk_topp_sampler.py:307]: Same here

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 4]
```diff
@@ -255,6 +255,14 @@ def apply_top_k_top_p(
         # Avoid sorting vocab for top-k only case.
         return apply_top_k_only(logits, k)
 
+    if p is not None and k is not None:
+        # Use optimized path only when max_k is small enough to benefit.
```
Comment (yewentao256) [inline vllm/v1/sample/ops/topk_topp_sampler.py:259]:
```suggestion
        # Use optimized path only when max_k < vocab_size
```

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work! A few thoughts

**Comment by njhill** [comment]
Thanks @nadavrot @yewentao256, yes I think we want to avoid `item()` now with async scheduling, so I don't think `torch.topk` is a good option anymore, even for the top_k-only case. I'm actually testing an alternative approach, will share later today.

**Comment by njhill** [comment]
Please see https://github.com/vllm-project/vllm/pull/32558!
